stdlib/asm/crypto/dsa/rpo_falcon512.masm

```diff
@@ -125,275 +125,185 @@ end
+    push.0.0
+    loc_storew.4
+
+    ## e) Store the random evaluation point. We will later check that it is the correct one.
+    movup.2 drop
+    movup.2 drop
+    loc_storew.8
+    # => [Y, Y, 0, 0, 0, 0, ptr, alpha_inv, acc1, acc0, ...]
+
+    # 2) Load the coefficients of the h polynomial and evaluate them at tau_inv
     repeat.64
         adv_pipe
 
+        # check that all coefficients are less than Falcon prime
         dupw.1
         u32assert2
-        push.M u32lt assert
-        push.M u32lt assert
+        u32overflowing_sub.M assert drop
+        u32overflowing_sub.M assert drop
         u32assert2
-        push.M u32lt assert
-        push.M u32lt assert
+        u32overflowing_sub.M assert drop
+        u32overflowing_sub.M assert drop
 
         dupw
         u32assert2
-        push.M u32lt assert
-        push.M u32lt assert
+        u32overflowing_sub.M assert drop
+        u32overflowing_sub.M assert drop
         u32assert2
-        push.M u32lt assert
-        push.M u32lt assert
+        u32overflowing_sub.M assert drop
+        u32overflowing_sub.M assert drop
+
+        horner_eval_base
 
         hperm
     end
 
-    # 3) Load saved claimed hash of h and compare
+    # 3) Load PK, the saved claimed hash of h, and compare
     loc_loadw.0
-    movup.4 assert_eq
-    movup.3 assert_eq
-    movup.2 assert_eq
-    assert_eq
-
-    # 4) Load s2 (Due to the final norm test we do not need to range check the s2 coefficients)
-    padw padw
-    repeat.64
-        adv_pipe hperm
-    end
-
-    # 5) Load claimed h * s2 in Z_Q[x]
-    repeat.128
-        adv_pipe hperm
-    end
-
-    # 6) Return the challenge point and the incremented pointer
-    exec.rpo::squeeze_digest
-    drop drop
-    #=> [tau1, tau0, ptr + 512*4]
-end
-
-#! Checks that pi == h * s2 in Z_Q[x] by evaluating both sides at a random point.
-#! The procedure takes as input a pointer h_ptr to h. The other two polynomials
-#! are located at h_ptr + 512, for s2, and h_ptr + 1024, for pi. The procedure takes
-#! also a pointer zeros_ptr to a region of memory [zeros_ptr, zeros_ptr + 512*4)
-#! and a pointer tau_ptr to powers of the random point we are evaluating at stored
-#! as [a_i, b_i, x, x] where (a_i, b_i) := tau^i for i in [0, 1023].
-#! The procedure returns () if the check passes, otherwise it raises an exception
-#! related to an unsatisfied assertion.
-#!
-#! Input: [h_ptr, zeros_ptr, tau_ptr, ...]
-#! Output: [...]
-#!
-#! Cycles: 2504
-export.probabilistic_product.16
-    # 1) Save the pointers
-    push.0 movdn.3
-    loc_storew.0
-
-    # 2) Compute the evaluation of the h polynomial at the random challenge
+    assert_eqw
+    # => [C, ptr, alpha_inv, h(tau)_1, h(tau)_0, ...]
 
-    # Accumulator to compute h(tau)
-    padw
-
-    # For mem_stream
+    # 4) Load s2 and evaluate at tau_inv
+    
+    ## a) Reset the accumulator, update the pointers and set up the state of the hasher
+    push.0 movdn.6
+    push.0 movdn.6
     padw padw
```

Comment:
Could we add stack state after this instruction?

Reply:
done